src/main/scala/com/actian/spark_vector/datastream/reader/RowReader.scala

```diff
@@ -0,0 +1,139 @@
+  }
+
+  private def fillColumnBuffers(vector: ByteBuffer) = {
+    numTuples = vector.getInt()
+    vector.order(ByteOrder.LITTLE_ENDIAN) /** The data from Vector comes in LITTLE_ENDIAN */
+    columnBufs.foreach { cb =>
+      cb.clear
+      cb.fill(vector, numTuples) /** Consume and deserialize tuples from the vector byte buffer */
+    }
+    vector.order(ByteOrder.BIG_ENDIAN) /** Go back to BIG_ENDIAN because this buffer is reused by DataStreamTap */
+    columnBufs
+  }
+
+  /**
+   * A list of calls (one per column buffer) to set typed values in the corresponding row's column position, and performing the necessary type casts
+   * while reading the needed value from its `ReadColumnBuffer[_]`.
+   *
+   * @note since read column buffers are exposed only through the `ReadColumnBuffer[_]` interface and we need to set the `SpecificMutableRow` with appropriate
+   * typed values, we make use of runtime reflection to determine the type of the generic parameter of the `ReadColumnBuffer[_]`
+   */
+  private def setValFromColumnBuffer(col: Int) = columnBufs(col).valueType match {
+    case y if y == classTag[Byte] => row.setByte(col, readValFromColumnBuffer[Byte](col))
+    case y if y == classTag[Short] => row.setShort(col, readValFromColumnBuffer[Short](col))
+    case y if y == classTag[Int] => row.setInt(col, readValFromColumnBuffer[Int](col))
+    case y if y == classTag[Long] => row.setLong(col, readValFromColumnBuffer[Long](col))
+    case y if y == classTag[Float] => row.setFloat(col, readValFromColumnBuffer[Float](col))
+    case y if y == classTag[Double] => row.setDouble(col, readValFromColumnBuffer[Double](col))
+    case y if y == classTag[BigDecimal] => row.update(col, Decimal(readValFromColumnBuffer[BigDecimal](col)))
+    case y if y == classTag[Boolean] => row.setBoolean(col, readValFromColumnBuffer[Boolean](col))
+    case y if y == classTag[UTF8String] => row.update(col, readValFromColumnBuffer[UTF8String](col))
+    case y => throw new Exception(s"Unexpected buffer column type '${y}'")
+  }
+
+  private def readValFromColumnBuffer[T: ClassTag](col: Int) = columnBufs(col).asInstanceOf[ReadColumnBuffer[T]].get()
+
+  private def read(): InternalRow = {
+    var col = 0
+    while (col < numColumns) {
+      setValAt(col)
+      col += 1
+    }
+    row
+  }
+
+  private def setValAt(col: Int) = if (columnBufs(col).isNextNull) {
+    row.setNullAt(col)
+  } else {
+    setValFromColumnBuffer(col)
+  }
+
+  override def hasNext(): Boolean = numTuples > 0 || {
+    profile("reading from datastream")
+    val ret = !tap.isEmpty
+    profileEnd
+    ret
+  }
+
+  override def next(): InternalRow = {
+    if (!hasNext) throw new NoSuchElementException("Empty row reader.")
+    if (numTuples == 0) {
+      val vector = tap.read()
+      profile("columns buffering")
+      fillColumnBuffers(vector)
+      profileEnd
+      // if hasNext read data we'll print two profiles here
+      profilePrint
```

Comment:
you don't want to print the profiles on each Vector either :) profilePrint should be called in close.